lib/src/client/client.dart

```diff
@@ -17,19 +17,60 @@ import 'dart:async';
 
 import 'call.dart';
 import 'channel.dart';
+import 'common.dart';
+import 'interceptor.dart';
 import 'method.dart';
 
 /// Base class for client stubs.
 class Client {
   final ClientChannel _channel;
   final CallOptions _options;
+  final List<ClientUnaryInterceptor> _unaryInterceptors;
+  final List<ClientStreamingInterceptor> _streamingInterceptors;
 
-  Client(this._channel, {CallOptions options})
-      : _options = options ?? CallOptions();
+  Client(this._channel,
+      {CallOptions options,
+      Iterable<ClientUnaryInterceptor> unaryInterceptors,
+      Iterable<ClientStreamingInterceptor> streamingInterceptors})
+      : _options = options ?? CallOptions(),
+        _unaryInterceptors =
+            List.unmodifiable(unaryInterceptors ?? Iterable.empty()),
+        _streamingInterceptors =
+            List.unmodifiable(streamingInterceptors ?? Iterable.empty());
 
   ClientCall<Q, R> $createCall<Q, R>(
       ClientMethod<Q, R> method, Stream<Q> requests,
       {CallOptions options}) {
     return _channel.createCall(method, requests, _options.mergedWith(options));
   }
+
+  ResponseFuture<R> $createUnaryCall<Q, R>(ClientMethod<Q, R> method, Q request,
+      {CallOptions options}) {
+    ClientUnaryInvoker invoker = (method, request, options) =>
+        ResponseFuture<R>(
+            _channel.createCall<Q, R>(method, Stream.value(request), options));
+
+    for (final interceptor in _unaryInterceptors.reversed) {
```

Comment:
Could this for-loop be in a separate helper function, like:

```
ClientInvoker invokeInterceptors(Iterable<ClientInterceptor> interceptors)
```

you would need to make the abstract ClientInterceptor class suggested in another comment.

Reply:
I was thinking about suggesting to abstract this away - but decided against it because Dart does not allow easy abstraction over method invocation. `invokeInterceptors` would need to take a closure which describes which of two methods `interceptUnary` or `interceptStreaming` to invoke so invoking `invokeInterceptors` would have to look like this: 

```dart
invokeInterceptors((interceptor, method, request, options, continuation) => 
    interceptor.interceptUnary(method, request, options, continuation));
```

This is just too much boilerplate.

Reply:
Ah dang. Thanks for the explanation -- still learning Dart nuances.

Reply:
I guess this is resolved then?

Reply:
Yep.